packages/notebook/src/tokens.ts

```diff
@@ -130,3 +137,62 @@ export interface INotebookTracker extends IWidgetTracker<NotebookPanel> {
    */
   readonly selectionChanged: ISignal<this, void>;
 }
+
+export interface IExecutionOptions {
+  /**
+   * Cell to execute
+   */
+  cell: Cell;
+  /**
+   * Notebook to which the cell belongs
+   */
+  notebook: Notebook;
+  /**
+   * A signal that emits whenever a cell completes execution.
+   */
+  executed: Signal<
+    any,
+    {
```

Comment:
1. Should `executed` and `executionScheduled` signals have an exported interface?
2. Should we make the full signal public or narrow it down to `emit` method?

Reply:
To clarify I am thinking something like `SignalEmitter = Pick<Signal, 'emit'>` (possibly not the right syntax but I hope you see what I mean)

Reply:
Another take: it seems that `Notebook` is mostly passed so that these signal can be emitted (`handlePayload` only needs notebook model really). Could we proxy emitting the signals and add the `notebook` outside to reduce how much we pass down in options?

I guess there is no getting around passing `Cell` object though.

Reply:
Thanks for the review @krassowski 

---

### Signal comments

The two signals `executed` and `executionScheduled` are [publicly available](https://github.com/jupyterlab/jupyterlab/blob/a0b9c86c93bbf66b070fb483ca2bc23beec671a6/packages/notebook/src/actions.tsx#L81) as `ISignal` from `NotebookActions`. So I don't see why we should drop `connect` and `disconnect`. Then we obviously need `emit`. So that left us only with the attribute `sender` that is exposed in `Signal` but not `ISignal` (that is a dummy placeholder in this case). So I would not bother creating a specific interface.

Another way to look at it, if those actions would have been implemented as Object method (never got why they are functions when most of the code is Object oriented), overriding the behavior would have meant inheriting from or duplicate `Notebook` with direct access to the signals (not the interface).

---

I'm fine passing the notebook model rather than the widget (it actually makes more sense). I think it is still important to pass the notebook document to get the context of execution. For example, one may require access to the notebook metadata (for the kernel information or custom info). Another scenario would be the reactive notebook case where cells depending on the current cell being executed should be queued for execution too.